morpheus/_lib/src/io/loaders/rest.cpp

```diff
@@ -17,21 +17,295 @@
+        http::write(stream, request);
+        beast::flat_buffer buffer;
+        http::read(stream, buffer, response);
+        beast::error_code ec;
+        stream.socket().shutdown(tcp::socket::shutdown_both, ec);
+        if (ec && ec != beast::errc::not_connected)
+        {
+            throw beast::system_error{ec};
+        }
+        int status = response.result_int();
+        // 503 Service Unavailable / 504 Gateway Timeout
+        if (status != 503 && status != 504)
+        {
+            break;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(interval_milliseconds));
+        interval_milliseconds *= 2;
+    }
+}
+
+void get_data_from_endpoint(http::response<http::dynamic_body>& response,
+                            const std::string& method,
+                            const std::string& endpoint,
+                            const std::string& params,
+                            const std::string& content_type,
+                            const std::string& body,
+                            const std::unordered_map<std::string, std::string>& x_headers,
+                            int max_retry)
+{
+    std::string ep(endpoint);
+    std::string query;
+    std::string host;
+    std::string target;
+    {
+        py::gil_scoped_acquire gil;
+
+        py::module_ urllib  = py::module::import("urllib.parse");
+        py::object ep_pystr = py::str(ep);
+
+        // Following the syntax specifications in RFC 1808, urlparse recognizes a netloc only if it is properly
+        // introduced by ‘//’. Otherwise the input is presumed to be a relative URL and thus to start with a path
+        // component. ref: https://docs.python.org/3/library/urllib.parse.html Workaround here is to prepend "//" if
+        // endpoint does not already include one
+        if (ep_pystr.attr("find")("//").cast<int>() == -1)
+        {
+            ep = "//" + ep;
+        }
+
+        py::object result = urllib.attr("urlparse")(ep);
+        host              = result.attr("hostname").is_none() ? "" : result.attr("hostname").cast<std::string>();
+        target            = result.attr("path").is_none() ? "" : result.attr("path").cast<std::string>();
+        if (target.empty())
+        {
+            target = "/";
+        }
+        query = result.attr("query").is_none() ? "" : result.attr("query").cast<std::string>();
+
+        // if params exists, overrides query included in endpoint
+        if (!params.empty())
+        {
+            query = params;
+        }
+    }
+
+    net::io_context ioc;
+    tcp::resolver resolver(ioc);
+    beast::tcp_stream stream(ioc);
+    try
+    {
+        auto const endpoint_results = resolver.resolve(host, PORT);
+        auto verb                   = get_http_verb(method);
+        auto request = construct_request(verb, target, query, host, method, content_type, body, x_headers);
+
+        try_get_data(stream, endpoint_results, request, response, max_retry, 1000);
```

Comment:
I am not extremely familiar with beast, so I don't know if it does url-encoding by default; but, since you're using urllib, it may be worth thinking a bit about these issues:

https://docs.python.org/3/library/urllib.parse.html#url-parsing-security
https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode